libass/ass_androidfonts.c

```diff
@@ -0,0 +1,659 @@
+            xskipwhile(strchr(XML_BLANKS, ch));
+            if (ch != '"' && ch != '\'')
+                xerror("Expected quotes for attribute value to follow");
+            char quote = ch;
+            i = 0;
+            while (i < sizeof(ret->attr.value)) {
+                xgetchar(&ch);
+                if (ch == quote)
+                    break;
+                ret->attr.value[i++] = ch;
+            }
+            if (i == sizeof(ret->attr.value))
+                xerror("Insufficient buffer space");
+            ret->attr.value[i] = 0;
+            return 1;
+        }
+        case XML_STATE_DOCEND: {
+            ret->type = XML_TYPE_EOF;
+            ret->flags = 0;
+            return 1;
+        }
+    }
+    xerror("State machine broke");
+}
+
+static int xml_next(struct xml_parser *p, struct xml_thing *ret)
+{
+    int r;
+    do {
+        r = xml_next2(p, ret);
+    } while (r == 0);
+    if (r == -1)
+        return r;
+
+    if (ret->type == XML_TYPE_TEXT_CONTENT) {
+        int useful = 0;
+        for (int i = 0; ret->text.content[i]; i++)
+            useful |= !strchr(XML_BLANKS, ret->text.content[i]);
+        if (!useful) // skip text nodes that are blank
+            return xml_next(p, ret);
+    } else if (ret->type == XML_TYPE_TAG) {
+        if (ret->tag.name[0]) {
+            if (ret->flags & XML_FLAG_OPENING_TAG) {
+                if (p->nested_tag_count == XML_NEST_DEPTH)
+                    xerror("Insufficient array space");
+                int cnt = p->nested_tag_count++;
+                strncpy(p->nested_tags[cnt], ret->tag.name, sizeof(*p->nested_tags));
+            } else if (ret->flags & XML_FLAG_CLOSING_TAG) {
+                if (p->nested_tag_count == 0)
+                    xerror("Superfluous closing tag");
+                const char* expect = p->nested_tags[--p->nested_tag_count];
+                if (strcmp(expect, ret->tag.name))
+                    xerror("Mismatching closing tag");
+            }
+        } else {
+            if (ret->flags & XML_FLAG_CLOSING_TAG) // self-closing
+                p->nested_tag_count--;
+        }
+    } else if (ret->type == XML_TYPE_EOF) {
+        if (p->nested_tag_count > 0)
+            xerror("Missing closing tag(s)");
+    }
+    return 1;
+}
+
+#undef xerror
+#undef xread
+#undef xgetchar
+#undef xskipwhile
+#undef xcopyloop
+
+static const char *xml_error(struct xml_parser *p)
+{
+    return p->error;
+}
+
+static void xml_free(struct xml_parser *p)
+{
+    free(p);
+}
+
+/*****************************************/
+
+#define FONTS_XML "/system/etc/fonts.xml"
+#define FONTS_PATH "/system/fonts"
+#define MAX_FULLNAME 100
+
+#define ALOGE(fmt, ...) __android_log_print(ANDROID_LOG_ERROR,   "libass", fmt, ##__VA_ARGS__)
+#define ALOGV(fmt, ...) __android_log_print(ANDROID_LOG_VERBOSE, "libass", fmt, ##__VA_ARGS__)
+#define ALOGD(fmt, ...) __android_log_print(ANDROID_LOG_DEBUG,   "libass", fmt, ##__VA_ARGS__)
```

Comment:
I added these just for debugging, libass has its own proper logging system afaik